### ISSUE TITLE
on startup, P3 sketch recognizes mouse but not keyboard (intermittent)

### ISSUE THREAD

**Issue by ottenm** [processing/processing#5359]
When I launch a sketch from the P3 IDE, sometimes the keyboard works directly and sometimes I have to click in the window first.  It's maybe 50/50, seems lower.  mouseMoved(), and mouseX/mouseY always work directly, not requiring a click in the window.  It's clear from the (Windows) taskbar and the title bar on the app, the sketch is the top app and has the focus.  I have a few dozen sketches.  The problem seems intermittent on all of them (not unique to this or that sketch).

No luck adding this to setup(): surface.setAlwaysOnTop(true);

I added this to draw()
```
if (KeyboardFocusManager.getCurrentKeyboardFocusManager() != null)
    println(millis() + ": " + KeyboardFocusManager.getCurrentKeyboardFocusManager().getFocusOwner());
```
The keystrokes fail when the focus owner shows JFrame or null, and the keystrokes work when the focus starts on the JFrame but then switches to processing.awt.PSurfaceAWT:$SmoothCanvas (either on it's own, *preferable*, or as the result of me clicking in the window).

Using 64-bit Windows 10 Pro and Processing IDE 3.3.

Any help/suggestions greatly appreciated!

**Comment by benfry** [comment]
The canvas requests focus on startup, but because that can only be requested, the OS may be taking it back or doing something else in the meantime. Sounds like it's happening out of sequence with the various threads that are at work (which is what usually causes intermittent issues like this).

**Comment by ottenm** [comment]
The focus does stay in the app (mouseX, mouseY, and mouseMoved() all work) but it definitely gets shuffled.  As a hack, is there a way to request it back to the surface/canvas ('surface' doesn't publish requestFocus()).  I'm thinking something like this in draw(): if (millis() < 1000) surface.requestFocus();

**Comment by benfry** [comment]
Yeah, from what you're seeing, it seems like mouse focus is handled separately. 

You can get the `Canvas` with `surface.getNative()` and request the focus on that (after a cast).

**Comment by ottenm** [comment]
A thousand thanks!!  Keyboard works every time!
